app/models/generic_event.rb

```diff
@@ -143,14 +143,43 @@ def self.details_attributes(*attributes)
     end
   end
 
+  # NB: Majority of events will use this serializer rather than the anonymous class.
+  def self.serializer
+    GenericEventSerializer
+  end
+
   # Relationship attributes live against the details but are expected in the json:api relationship section
   # so are defined separately
   def self.relationship_attributes(attributes)
     define_singleton_method(:relationship_attributes) do
       instance_variable_get('@relationship_attributes')
     end
+
     instance_variable_set('@relationship_attributes', attributes)
 
+    define_singleton_method(:serializer) do
+      @serializer ||=
+        begin
+          serializer = Class.new(GenericEventSerializer)
+
+          relationship_attributes.each do |attribute_key, attribute_type|
+            # NB: Default to V2 relationship serializer and fallback to unversioned serializer. Never support v1
+            relation_serializer = "V2::#{attribute_type.to_s.singularize.camelize}Serializer"
```

Comment:
Might be nice to use a separate variable for this rather than reassigning the cla